### ISSUE TITLE
Nested Router

### ISSUE THREAD

**Issue by TiagoBraga39** [angular/angular#3751]
Hi all!

I have been fooling around with routers, and it seems to me that a nested router can only navigate to a inner Route, through the parent router.
A simple example with alpha 35:

---

<b>Outer Router:</b>
```javascript
@RouteConfig([
    //new Redirect({ path: '/inner', redirectTo: 'inner/favorites' }), // Issue 2
    new Route({ path: '/inner/...', component: InnerRouter, as: 'inner' })
])
```

<b>Inner Router:</b>
```javascript
@RouteConfig([
    //new Redirect({ path: '/', redirectTo: '/favorites' }), // Issue 3
    new Route({ path: ':path', component: Component, as: 'path' })
])
```

---

<b>Issue 1:</b>
If I currently am at the inner router, why can't I navigate within it (to a path located at the inner router) by using: "navigate('path')"? I always need to "nagivate('nested/path')".

<b>Issue 2 </b>: (check the commented line on outer router)
If I use that redirect in the outer router, the path will be: 'inner/favorites/path'.

<b>Issue 3 </b>: (check the commented line on inner router)
If I use that redirect in the inner router, I won't be able to simply travel to inner router with: navigate('/inner'). Inner router path '/' does not resolve with redirect.

Best regards,
Tiago Braga

**Comment by btford** [comment]
1. Right now, `navigate` only accepts the entire URL. I've been meaning to rename `navigate` to `navigateUrl`, and introduce a new method, `navigate` that uses the same syntax as the `RouterLink` directive, but haven't gotten around to it yet. For now, you can do `router.navigateInstruction(router.generate(['./foo']))`.

2/3. These need some investigation. Thanks for the bug report!

**Comment by TiagoBraga39** [comment]
Thank you for all the effort you guys have put to the development of the best framework ever!
Keep it up! +1

**Comment by TiagoBraga39** [comment]
@btford Hi again.

---

Just a thought that came up:
I can't find this on the source code, but I am not saying it is not there.
<b>When routing on a Child Router, if the route if not found, what is the behaviour?</b>

Does the routing process try to resolve the request in the Parent Router, GrandParent and so on?
If so, when does it trigger the Redirect? I guess it should be the first Redirect hit, <b>if</b> there is not a router that can resolve the request.

Best regards,
Tiago Braga

**Comment by hitmanmcc** [comment]
@TiagoBraga39 Good question, I'm actually curious about the response to that. I'm wondering if it works in a similar fashion to DI and if it's subject to control mechanisms like regular DI is, i.e., if you can tell a router how far it can go up the tree to look for a route.

**Comment by btford** [comment]
@TiagoBraga39 What do you think of using the routing DSL as the value of the `redirectTo` property?

It's less ambiguous, and handles cases like your number 3 much more elegantly.

**Comment by TiagoBraga39** [comment]
I am not sure I understood the question, but I guess the easier way to redirect would be this.
Again the router:
```javascript
@RouteConfig([
    new Redirect({ path: '/', redirectTo: [ 'path_alias', { param1: value1, param2: value2 } ] }), 
    new Route({ path: ':param1/:param2', component: Component, as: 'path_alias' })
])
```
It is very similar to what router-link works.

By the way, is it possible to have an "Otherwise" class to route to 404 pages?

Best regards,
Tiago Braga

**Comment by btford** [comment]
@TiagoBraga39 

A quick update:

1. Fixed in d9036c6cf37e761306ca19931132f5af6c0ac0ed
2/3. I've opened #4170 as a more general fix for redirects.

I'm going to close this, but once #4170 is addressed, feel free to open new issues if the cases you describe still do not work.

Thanks!

**Comment by TiagoBraga39** [comment]
@btford Will do!
Thank you!
